### PR TITLE
chore: cpp framework improvements and additions

### PR DESCRIPTION
**Pull request by frankkopp**
## Summary of Changes
General additions and improvements to the cpp framework. 

- added arinc library
- added Google Test GTest framework
- added unit tests for cpp libraries
- refactored InterialDampener to a general DampingController
- additional clang format to better align code for multiple assignment statements
- gitignore cleanup

Discord username (if different from GitHub): cdr_maverick

## Testing instructions
Potentially Pushback, Lighting Presets, Aircraft Presets might be impacted by these changes. 
Mainly Pushback (b/o the DampingController refactor).
Other side effects unlikely.

A32NX: see above - test for regression
A380X: see above - test for regression

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

### CODE REVIEW

File: fbw-a32nx/src/wasm/extra-backend-a32nx/src/Gauge_Extra_Backend.cpp [hunk 1]
```diff
@@ -49,7 +49,7 @@ AircraftPresets aircraftPresets(msfsHandler, AircraftPresetProcedures_A32NX::air
  * https://docs.flightsimulator.com/html/Content_Configuration/SimObjects/Aircraft_SimO/Instruments/C_C++_Gauges.htm?rhhlterm=_gauge_callback&rhsearch=_gauge_callback
  */
 extern "C" {
-[[maybe_unused]] MSFS_CALLBACK bool Gauge_Extra_Backend_gauge_callback([[maybe_unused]] FsContext ctx, int svcId, void* pData) {
+MSFS_CALLBACK bool Gauge_Extra_Backend_gauge_callback([[maybe_unused]] FsContext ctx, int svcId, void* pData) {
```
Comment (2hwk) [inline fbw-a32nx/src/wasm/extra-backend-a32nx/src/Gauge_Extra_Backend.cpp:52]: [[maybe_unused]] 

Can this be cleaned up at this moment? Or for now are we still uncertain.

File: package.json [hunk 2]
```diff
@@ -6,24 +6,20 @@
     "#build-utils": "./build-utils.js"
```
Comment (2hwk) [inline package.json:6]: Intentional diff here?

**Review by 2hwk — CHANGES_REQUESTED** [review]

**Comment by 2hwk** [comment]
Just some early comments, looking good so far.

File: fbw-a32nx/src/wasm/extra-backend-a32nx/src/Gauge_Extra_Backend.cpp (hunk 1 above)
Comment (frankkopp) [reply fbw-a32nx/src/wasm/extra-backend-a32nx/src/Gauge_Extra_Backend.cpp:52]: I typically let my IDE guide me on this. 

Previously it showed a warning for the whole function which it does not any longer - probably was a change in the compiler. There were never any warnings when compiled via the dev-env for the first `[[maybe_unused]]` - the one I removed.

The second `[[maybe_unused]]`is necessary as indeed the ctx is not used:
image

I could alternatively just remove the var name which basically means the same thing as `[[maybe_unused]]`:
```cpp
MSFS_CALLBACK bool Gauge_Extra_Backend_gauge_callback(FsContext, int svcId, void* pData) {
```

If you prefer the second let me know - doesn't matter either way for me. 
It is just a bit clearer with the `[[maybe_unused]]`that the var is intentionally not used.

File: package.json (hunk 2 above)
Comment (frankkopp) [reply package.json:6]: If you mean the whole file then yes.
Added this:
```
"build:cpp-wasm-cmake-test": "cd fbw-common/src/wasm/cpp-msfs-framework/test && ./test-cpp.sh",
```
and file autoformat (json) removed the empty lines which is why I extended the divider lines to find stuff easier. 

The line you marked does not have any changes, does it?

Comment (2hwk) [reply package.json:6]: Yes I meant the whole file, thanks for the response.

File: fbw-a32nx/src/wasm/extra-backend-a32nx/src/Gauge_Extra_Backend.cpp (hunk 1 above)
Comment (2hwk) [reply fbw-a32nx/src/wasm/extra-backend-a32nx/src/Gauge_Extra_Backend.cpp:52]: Ok, this is fine, as long as it is intentional.

**Review by 2hwk — APPROVED** [review]

**Comment by Popespice** [comment]
Quality Assurance Tester Report

Discord          : popespice
Object of testing: 8640
Tier of Testing  : 2
Date             : 21/8/2024

Testing Process: 
1. Start sim as usual
2. Load into A32NX
3. Power up plane as normal
4. Individually select different presets (lighting and aircraft)
5. Taxi aircraft through built in taxi mechanism in EFB
6. Observe all presets work, aircraft flies as expected.

Negatives:       
N/A
Testing Results: 
Passed
Conclusions: 
Working as expected.

**Comment by tylerclt** [comment]
Quality Assurance Trainee Report

Discord: Tylerb780#3594
Object of testing: #8640 
Tier of Testing: 2
Date: 22/08/2024)

Testing Process: 
1. Loaded into the Aircraft
2. Powered-up aircraft as normal 
3. Went through each individual lighting and aircraft preset
4.  Pushed using the built-in pushback system
5.  Made sure aircraft flew with no issues or abnormalities


Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
Working as intended

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Table1502_A380X.hpp [hunk 3]
```diff
@@ -30,23 +30,22 @@ class Table1502_A380X {
    * @return A 2D array representing the CN2 - correctedN1 pairs.
    */
   static constexpr double table1502[13][4] = {
```
Comment (Maximilian-Reuter) [inline fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Table1502_A380X.hpp:32]: The values in the comments here dont reflect the table values and seem copied from the a32nx

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/ThrustLimits_A380X.hpp [hunk 4]
```diff
@@ -37,82 +37,82 @@ class ThrustLimits_A380X {
...
+      {9000,    15.000,  36.000,  94.031, 86.889, 60.968}, //
+      {10000,   13.000,  34.000,  94.957, 88.044, 60.935}, //
+      {11000,   12.000,  32.000,  95.295, 88.526, 59.955}, //
+      {12000,   11.000,  30.000,  95.568, 88.818, 58.677}, //
+      {13000,   10.000,  28.000,  95.355, 88.819, 59.323}, //
+      {14000,   10.000,  26.000,  95.372, 89.311, 59.965}, //
+      {15000,   8.000,   24.000,  95.686, 89.907, 58.723}, //
+      {16000,   5.000,   22.000,  96.160, 89.816, 57.189}, //
+      {16600,   5.000,   22.000,  96.560, 89.816, 57.189}, //
+      {-2000,   47.751,  54.681,  84.117, 81.901, 63.498}, // row 20
```
Comment (Maximilian-Reuter) [inline fbw-a380x/src/wasm/fadec_a380x/src/Fadec/ThrustLimits_A380X.hpp:62]: this line seems to belong to the GA part of the table and not the TO part

Comment (frankkopp) [reply fbw-a380x/src/wasm/fadec_a380x/src/Fadec/ThrustLimits_A380X.hpp:62]: > this line seems to belong to the GA part of the table and not the TO part

I did not touch the values (at least  not on purpose) - these should be the same as before (from the FADEC rewrite where I also did not touch values. 

The changes to this file are only formatting in this PR.

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Table1502_A380X.hpp (hunk 3 above)
Comment (frankkopp) [reply fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Table1502_A380X.hpp:32]: see above - did not touch the values

**Comment by pilotseyea350** [comment]
Quality Assurance Trainee Report

Discord: piloteyesA350
Object of testing: #8640
Tier of Testing: 2
Date: 23/08/2024

Testing Process:

- Aircraft Loaded at EDDM
- Power Up using Aircraft States 
- Set different cockpit lights configuration and save them
- Pushback using the EFB Pushback
- Fly from EDDM to EDDS
- Monitor systems
- Have fun
Negatives:


- "Ready to pushback" state completion is reliant on ADIRS alignment time which, if set to Real, does not happen until either ADIRS Alignment time is set to "Instant" or ADIRS is aligned. Known Issue but Fast Loading should override EFB Realism settings

- As discussed in QA Channel, Transition from "Ready to Pushback" to "Ready To Taxi" triggers Master Warning & Master Caution due to Flaps and Slats not being in TO Position

- Saving a lightning preset and not renaming it prevents it from being used as an automatic loaded Lightning preset. We could also change the Default names of the presets to "No Name 1" for preset 1, "No Name 2" for preset 2, etc...

- CKPT, FWD CAB & AFT CAB DUCT OVHT ( Ecam message after switching from an Engines running preset to a engines non running preset)

Testing Results:
Not Passed

Conclusions:
ADIRS Issue was known but Master Caution & Master Warning is not expected.

N. B. : 
#8581 
https://streamable.com/vkq57o

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/ThrustLimits_A380X.hpp (hunk 4 above)
Comment (Maximilian-Reuter) [reply fbw-a380x/src/wasm/fadec_a380x/src/Fadec/ThrustLimits_A380X.hpp:62]: I know I didnt mean to imply it had, just that this is smth that I noticed recently so wondered if you maybe want to quickly change that

Comment (frankkopp) [reply fbw-a380x/src/wasm/fadec_a380x/src/Fadec/ThrustLimits_A380X.hpp:62]: Ah ok - no problem. If you want you can also just push to the PR. I'm on vacation with no PC.

**Comment by frankkopp** [comment]
>   Negatives:
> * "Ready to pushback" state completion is reliant on ADIRS alignment time which, if set to Real, does not happen until either ADIRS Alignment time is set to "Instant" or ADIRS is aligned. Known Issue but Fast Loading should override EFB Realism settings

This is known and intendent for the current implementation - it is improved in https://github.com/flybywiresim/aircraft/pull/8581 as it required system internal changes. 

> * As discussed in QA Channel, Transition from "Ready to Pushback" to "Ready To Taxi" triggers Master Warning & Master Caution due to Flaps and Slats not being in TO Position

Not an issue caused by this PR (already known issue) - but workaround has been put in place. 
Please retest this workaround (it basically waits until the Flaps/Slats are deployed before pressing T.O CONFIG)

This is improved/fixed in https://github.com/flybywiresim/aircraft/pull/8581 as it required system internal changes. 

> * Saving a lightning preset and not renaming it prevents it from being used as an automatic loaded Lightning preset. We could also change the Default names of the presets to "No Name 1" for preset 1, "No Name 2" for preset 2, etc...

How is this prevented? Could not reproduce it. 
(Likely also outside this PR)

> * CKPT, FWD CAB & AFT CAB DUCT OVHT ( Ecam message after switching from an Engines running preset to a engines non running preset)

Could not be reproduced - please re-test with latest PR build.

> 
> Testing Results: Not Passed
> 
> Conclusions: ADIRS Issue was known but Master Caution & Master Warning is not expected.

**Comment by alepouna** [comment]
@frankkopp Is this still WIP/are you doing enhancements to this PR? If so I recommend to remove the QA RTT label. If you are doing improvements and need tests OR it's ready, feel free to notify QA on Discord as well 👌

**Comment by frankkopp** [comment]
> @frankkopp Is this still WIP/are you doing enhancements to this PR? If so I recommend to remove the QA RTT label. If you are doing improvements and need tests OR it's ready, feel free to notify QA on Discord as well 👌 

This is ready but didn't get any reply from the last QA tester after my comment to the test report.

**Comment by pilotseyea350** [comment]
Quality Assurance Trainee Report

Discord Username  : PilotEyesA350 
Object of testing : #8640  
Aircraft          : A32NX 
Tier of Testing   : 1
Date              : 03/09/2024 

Testing Process: 
- Spawn at Gate
- Load "Ready to Taxi" Preset
- Wait until A/C is ready to use
- Load "Ready for Pushback" Preset

Testing Results: 
Not Passed

Negatives:       
Display Brightness are turned completely off when loading "Ready to Taxi" Preset

Conclusions: 
CAB DUCT OVHT Has been Fixed, Same for ADIRS Alignment, Same for MASTER WARN & CAUT when loading "Ready for Takeoff"

Test not passed as the brightness prevent the user to use the aircraft as soon as a preset is loaded. An additionnal Tier 2 test will be performed when new fixes will be introduced.

Lightning preset bug is shown in the media section

Media:
Brightness :
https://streamable.com/xpspfo

Lightning preset :
https://streamable.com/tuiudk

**Comment by frankkopp** [comment]
> Media: Brightness : [streamable.com/xpspfo](https://streamable.com/xpspfo)

This is not an issue of this PR and rather seems to be a faulty setup of the auto setting for Lighting Presets.
Turn this off and redo the test then it should not happen. Also see below.  
 
> Lightning preset : [streamable.com/tuiudk](https://streamable.com/tuiudk)

This is also not an issue of the PR - if this is indeed an issue it has most likely been existing previously. 
I recommend to create a separate issue so it can be fixed in a separate PR. 
This will not be fixed by this PR

If you could check both things with the current development version - I bet both issues already exist without this PR.

**Comment by alepouna** [comment]
to add to the above, I recommend also trying a clean work folder by renaming the A32NX folder to try and replicate.

**Comment by pilotseyea350** [comment]
> > Media: Brightness : [streamable.com/xpspfo](https://streamable.com/xpspfo)
> 
> This is not an issue of this PR and rather seems to be a faulty setup of the auto setting for Lighting Presets. Turn this off and redo the test then it should not happen. Also see below. image
> 
> > Lightning preset : [streamable.com/tuiudk](https://streamable.com/tuiudk)
> 
> This is also not an issue of the PR - if this is indeed an issue it has most likely been existing previously. I recommend to create a separate issue so it can be fixed in a separate PR. This will not be fixed by this PR
> 
> If you could check both things with the current development version - I bet both issues already exist without this PR.

Understood, I'll try this ASAP

**Comment by pilotseyea350** [comment]
> > Media: Brightness : [streamable.com/xpspfo](https://streamable.com/xpspfo)
> 
> This is not an issue of this PR and rather seems to be a faulty setup of the auto setting for Lighting Presets. Turn this off and redo the test then it should not happen. Also see below. image
> 
> > Lightning preset : [streamable.com/tuiudk](https://streamable.com/tuiudk)
> 
> This is also not an issue of the PR - if this is indeed an issue it has most likely been existing previously. I recommend to create a separate issue so it can be fixed in a separate PR. This will not be fixed by this PR
> 
> If you could check both things with the current development version - I bet both issues already exist without this PR.

So the lightning preset issue is indeed an issue with the current dev version. Now for the screen brightness issue, I could not reproduce it with the dev build even with a clean work folder.

**Comment by frankkopp** [comment]
> > > Media: Brightness : [streamable.com/xpspfo](https://streamable.com/xpspfo)
> > 
> > 
> > This is not an issue of this PR and rather seems to be a faulty setup of the auto setting for Lighting Presets. Turn this off and redo the test then it should not happen. Also see below. image
> > > Lightning preset : [streamable.com/tuiudk](https://streamable.com/tuiudk)
> > 
> > 
> > This is also not an issue of the PR - if this is indeed an issue it has most likely been existing previously. I recommend to create a separate issue so it can be fixed in a separate PR. This will not be fixed by this PR
> > If you could check both things with the current development version - I bet both issues already exist without this PR.
> 
> So the lightning preset issue is indeed an issue with the current dev version. Now for the screen brightness issue, I could not reproduce it with the dev build even with a clean work folder.

Did you test with this turned off?
image

**Comment by pilotseyea350** [comment]
Using this pr's build ?

**Comment by frankkopp** [comment]
> Using this pr's build ?

Yes - as I believe this comes from the Auto-preset - if not I would not have an idea just now why this happens for you. It never happened to me or any other tester. Also I couldn't think of a circumstance that my changes could cause this.

**Comment by pilotseyea350** [comment]
I'll try again this afternoon with the latest build

**Comment by pilotseyea350** [comment]
Just tested it and it did work with the auto-preset off. Will do a T2 asap toalow you to merge

**Comment by pilotseyea350** [comment]
Capture d'écran 2024-09-11 220459
Is the vertical profile green dot below the selected altitude expected ?

**Comment by frankkopp** [comment]
> Capture d'écran 2024-09-11 220459
> 
> Is the vertical profile green dot below the selected altitude expected ?

The cpp code of this PR does not touch any of this. Its really only used in the backend (not front end / UI) for Lighting Preset, Aircraft Preset, pushback, FADEC. 
Frontend is in Typescript which is why the Automatic Lighting presets is also not touched as this logic is TS only.

**Comment by tracernz** [comment]
And it's also perfectly fine, so no issue 👍 .

**Comment by pilotseyea350** [comment]
Quality Assurance Trainee Report

Discord Username  : PilotEyesA350 
Object of testing : #8640  
Aircraft          : A32NX
Tier of Testing   : 2
Date              : 11/09/2024 

Testing Process: 
- Load aircraft at gate
- Follow A320 flows
- Fly from EDDF to EDDF using a Simbrief Flightplan
- Monitor systems

Testing Results: 
Passed  

Negatives:       
N/A

Conclusions: 
After much work it finally works as intended ! 

Media:
N/A

**Comment by frankkopp** [comment]
> Quality Assurance Trainee Report
> 
> Discord Username : PilotEyesA350 Object of testing : #8640 Aircraft : A32NX Tier of Testing : 2 Date : 11/09/2024
> 
> Testing Process:
> 
> * Load aircraft at gate
> * Follow A320 flows
> * Fly from EDDF to EDDF using a Simbrief Flightplan
> * Monitor systems
> 
> Testing Results: Passed
> 
> Negatives: N/A
> 
> Conclusions: After much work it finally works as intended !
> 
> Media: N/A

Thanks - your effort is very much appreciated!!

**Pull request merged**